### PR TITLE
report: add fractional category scores

### PR DESCRIPTION
**Pull request by adamraine**
This moves the category ratio indicator from the flow report down to the legacy report renderer. ~There are still some style details to work out, but I wanted to get some initial impressions on the structure.~

Screen Shot 2021-09-08 at 3 34 07 PM

### CODE REVIEW

**Comment by adamraine** [comment]
I'm starting to think that we should use a score gauge for timespan performance and snapshot accessibility. Quick mock:

Screen Shot 2021-09-07 at 2 27 42 PM

Snapshot accessibility makes the most sense because it's the same 54 audits that are found in navigation mode, so it makes sense to represent it the same way.

For timespan performance, it could be confusing when the category is graded as "average" even though the metrics are all passing ([example](https://lighthouse-74nif1ib1-googlechrome.vercel.app/sample-reports/flow-report/#index=1&anchor=performance)). The unscored diagnostics and opportunities heavily impact how the category is scored, even though the metrics are the only audits scored. I think a score gauge makes more sense than a ratio in this case.

**Comment by adamraine** [comment]
It would be a simple change if we do decide to make it, so I'm willing to defer to our previous position and finish this PR as is. I had a couple more thoughts if we decide to revisit: 

> Long-term I have hope for more metric-driven scores on timespans but we really need to let the space mature and establish some benchmarks for success before we can confidently provide a holistic score in this area, IMO.

I agree that a 0-100 score for just TBT and CLS isn't ideal, I just thought it might be better than the fraction. I guess I'm hoping for a more metric-driven score as well :)

> 1. Reinforcing "this isn't the same navigation environment" is more important than conveying the a11y score out of 0-100.
> 2. There's been a lot of pushback on a11y in particular for mistakenly conveying "100" when there's a lot not covered. Switching to a fraction helps alleviate that and remind folks it's not comprehensive.
>
> I think with the consistency argument plus these two points, I still lean toward the fraction.

The problem is that a11y *is* the same between the two environments, so I'm still worried about the consistency within the category.

File: report/assets/templates.html [hunk 1]
```diff
@@ -562,6 +562,18 @@
   </a>
 </template>
 
+<!-- Lighthouse category ratio -->
+<template id="ratio">
```
Comment (adamraine) [inline report/assets/templates.html:566]: I've been calling this "ratio", but I've seen others refer to it as "fraction". Any strong opinions about the name?

Comment (patrickhulce) [reply report/assets/templates.html:566]: I typically refer to a representation like this as a "fraction" when the first number is strictly less than or equal to the second number, and a "ratio" when the two numbers have a more arbitrary relationship.

By that lens, I would call these a fraction.

File: report/renderer/category-renderer.js [hunk 2]
```diff
@@ -317,6 +317,10 @@ export class CategoryRenderer {
    * @return {DocumentFragment}
    */
   renderScoreGauge(category, groupDefinitions) { // eslint-disable-line no-unused-vars
+    if (category.displayMode === 'ratio') {
```
Comment (patrickhulce) [inline report/renderer/category-renderer.js:320]: this seems like a bit of a misleading path where we've asked for a gauge but receive a ratio

new method `renderCategoryScore` that does this delegation like you've done in the flow report?

File: flow-report/src/common.tsx [hunk 3]
```diff
@@ -6,49 +6,8 @@
...
-  let rating = Util.calculateRating(ratio);
-
-  // If none of the available audits can affect the score, a rating isn't useful.
-  // The flow report should display the ratio with neutral icon and coloring in this case.
-  if (totalWeight === 0) {
-    rating = 'null';
-  }
-
-  return (
-    <a href={href} className={`CategoryRatio CategoryRatio--${rating}`} data-testid="CategoryRatio">
```
Comment (patrickhulce) [inline flow-report/src/common.tsx:46]: removed matching styles?

File: flow-report/src/wrappers/category-score.tsx [hunk 4]
```diff
@@ -9,17 +9,21 @@ import {useEffect, useLayoutEffect, useRef} from 'preact/hooks';
...
   const ref = useRef<HTMLDivElement>(null);
 
   useLayoutEffect(() => {
     const el = categoryRenderer.renderScoreGauge(category, {});
 
     // Category label is displayed in the navigation header.
-    const label = el.querySelector('.lh-gauge__label');
+    const label = el.querySelector('.lh-gauge__label,.lh-ratio__label');
     if (label) label.remove();
+    const bg = el.querySelector('.lh-ratio__background');
```
Comment (patrickhulce) [inline flow-report/src/wrappers/category-score.tsx:25]: comment for this one? 

my expectation is the summary doesn't have the background, but I still expect individual reports to have the background (but those are using the report renderer wrapper and not this component)

**Review by patrickhulce — APPROVED** [review]

File: report/renderer/category-renderer.js (hunk 2 above)
Comment (adamraine) [reply report/renderer/category-renderer.js:320]: I started doing this but backed out once I reached
https://github.com/GoogleChrome/lighthouse/blob/ee53af31fca149eeede3b9ac217d2b5882a58782/report/renderer/report-renderer.js#L167

We would just keep that line the same right?

Comment (patrickhulce) [reply report/renderer/category-renderer.js:320]: My preference would be to replace this line with `renderCategoryScore === renderCategoryScore` and rename the PWA one to renderCategoryScore. In fact that name makes more sense anyway because it doesn't have a gauge :)

We can alias `renderScoreGauge` to `renderCategoryScore` in PWA category for backcompat until v9 in case any partners are using it.

File: report/types/report-result.d.ts [hunk 5]
```diff
@@ -17,7 +17,8 @@ interface ReportResult extends LHResult {
 
 declare module ReportResult {
   interface Category extends LHResult.Category {
-    auditRefs: Array<AuditRef>
+    auditRefs: Array<AuditRef>;
+    displayMode: 'gauge'|'fraction';
```
Comment (brendankenny) [inline report/types/report-result.d.ts:21]: thinking of the API proposal in https://github.com/GoogleChrome/lighthouse/pull/12772/ (where `prepareReportResult` won't necessarily be used) and the possibility there might be more changes needed to category rendering besides just gauge vs fraction, what do you think about an options object passed into `renderCategoryScore`?

It seems like that's ultimately what we're going to want from `renderReport()`/`renderReportPartials()` and it would be nice to not alter `ReportResult` further from the LHR.

Comment (adamraine) [reply report/types/report-result.d.ts:21]: > what do you think about an options object passed into renderCategoryScore?

That would mean passing the options through multiple layers of function calls, in multiple different locations. We could do it, but it seems like `ReportResult` was designed so we don't have to make these kinds of structural changes.

Comment (brendankenny) [reply report/types/report-result.d.ts:21]: `ReportResult` wasn't so much designed :)

https://github.com/GoogleChrome/lighthouse/blob/5cbad246937b6332aca9be65780c98a20bc8926d/report/renderer/util.js#L55

It is a structural change and it would have to go through a few layers (though not so many). I think it may be worth it, though:
- I want to call `renderGaugeForCategory(category, groups, options)`, I don't want to have to call `prepareReportResult` first
- if there are other per-mode changes (e.g. if we go with #13022, perf is going to be different than the other categories in navigation reports but not in timespan/snapshot), we're going to have to expand `displayMode` or generalize it anyways

basically if we're going to parameterize on gatherMode, let's just do it outright rather than roundabout through `ReportResult` :)
